File: learnLeetCode/Day14/Part1/parse_maze.cpp
#include <bits/stdc++.h>
using namespace std;

// Part 1 — Parse the Maze
// Helper to find positions of a target character.

vector<pair<int,int>> findPositions(const vector<vector<char>>& maze, char target) {
    int n = (int)maze.size();
    vector<pair<int,int>> res;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < (int)maze[i].size(); ++j) {
            if (maze[i][j] == target) res.emplace_back(i, j);
        }
    }
    return res;
}

optional<pair<int,int>> findSingle(const vector<vector<char>>& maze, char target) {
    int n = (int)maze.size();
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < (int)maze[i].size(); ++j) {
            if (maze[i][j] == target) return make_pair(i, j);
        }
    }
    return nullopt;
}

int main() {
    vector<vector<char>> maze = {
        {'S','E','P','W'},
        {'E','W','E','E'},
        {'E','P','W','E'},
        {'E','E','E','X'}
    };

    auto ps = findPositions(maze, 'P');
    cout << "Positions of 'P': ";
    for (auto &p : ps) cout << "(" << p.first << "," << p.second << ") ";
    cout << "\n";

    auto s = findSingle(maze, 'S');
    auto x = findSingle(maze, 'X');
    cout << "Start 'S': ";
    if (s) cout << "(" << s->first << "," << s->second << ")\n"; else cout << "None\n";
    cout << "Exit  'X': ";
    if (x) cout << "(" << x->first << "," << x->second << ")\n"; else cout << "None\n";
    return 0;
}

File: learnLeetCode/Day14/Part1/parse_maze.py
from typing import List, Tuple, Optional

# Part 1 — Parse the Maze
# Write helper functions to locate 'S', 'X', and collect any target positions like 'P'.

Pos = Tuple[int, int]
Grid = List[List[str]]

def find_positions(maze: Grid, target: str) -> List[Pos]:
    """Return all (row, col) positions where maze[row][col] == target.

    Assumes maze is a rectangular list of lists of single-character strings.
    """
    n = len(maze)
    res: List[Pos] = []
    for i in range(n):
        row = maze[i]
        for j in range(len(row)):
            if row[j] == target:
                res.append((i, j))
    return res


def find_single(maze: Grid, target: str) -> Optional[Pos]:
    """Return the first position of target if present, else None."""
    n = len(maze)
    for i in range(n):
        row = maze[i]
        for j in range(len(row)):
            if row[j] == target:
                return (i, j)
    return None


if __name__ == "__main__":
    maze = [
        ['S','E','P','W'],
        ['E','W','E','E'],
        ['E','P','W','E'],
        ['E','E','E','X']
    ]
    print("Positions of 'P':", find_positions(maze, 'P'))
    print("Start 'S':", find_single(maze, 'S'))
    print("Exit  'X':", find_single(maze, 'X'))
